src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -1756,32 +1756,50 @@ public static decimal Sum<TSource>(this IEnumerable<TSource> source, Func<TSourc
         public static int Min(this IEnumerable<int> source)
         {
             if (source == null) throw Error.ArgumentNull("source");
-            int value = 0;
-            bool hasValue = false;
-            foreach (int x in source)
+            int value;
+            using (IEnumerator<int> e = source.GetEnumerator())
             {
-                if (hasValue)
+                if (!e.MoveNext()) throw Error.NoElements();
+                value = e.Current;
+                while (e.MoveNext())
                 {
+                    int x = e.Current;
                     if (x < value) value = x;
                 }
-                else
-                {
-                    value = x;
-                    hasValue = true;
-                }
             }
-            if (hasValue) return value;
-            throw Error.NoElements();
+            return value;
         }
 
         public static int? Min(this IEnumerable<int?> source)
         {
             if (source == null) throw Error.ArgumentNull("source");
-            int? value = null;
-            foreach (int? x in source)
+            int? value;
+            using (IEnumerator<int?> e = source.GetEnumerator())
             {
-                if (value == null || x < value)
-                    value = x;
+                if (!e.MoveNext()) return null;
+                value = e.Current;
```

Comment:
As I remember a regular while{} is emitted as a do{}while  with entry point branching to a label just before the exit condition. This way the looping code path has just one branch.

It is a bit surprising that while{] cost more. It could be something with branch prediction, or perhaps the "return null" sequence causes some effects. - It would need to zero-init a temp, and then copy it into a return temp (since we are in a try/finally), not sure if JIT can follow this. If it is all about "return null", then you may be able to squeeze some cycles by pre-initing value to null and then just goto to the "return value".